SignalServiceKit/src/Messages/Attachments/OWSAttachmentDownloads.m

```diff
@@ -35,42 +36,104 @@
 // indicator shows up as quickly as possible.
 static const CGFloat kAttachmentDownloadProgressTheta = 0.001f;
 
-@interface OWSAttachmentsProcessor ()
+typedef void (^AttachmentDownloadSuccess)(TSAttachmentStream *attachmentStream);
+typedef void (^AttachmentDownloadFailure)(NSError *error);
 
-@property (nonatomic, readonly) TSNetworkManager *networkManager;
+@interface OWSAttachmentDownloadJob : NSObject
+
+@property (nonatomic, readonly) TSAttachmentPointer *attachmentPointer;
+@property (nonatomic, readonly, nullable) TSMessage *message;
+@property (nonatomic, readonly) AttachmentDownloadSuccess success;
+@property (nonatomic, readonly) AttachmentDownloadFailure failure;
 
 @end
 
-@implementation OWSAttachmentsProcessor
+#pragma mark -
+
+@implementation OWSAttachmentDownloadJob
 
 - (instancetype)initWithAttachmentPointer:(TSAttachmentPointer *)attachmentPointer
-                           networkManager:(TSNetworkManager *)networkManager
+                                  message:(nullable TSMessage *)message
+                                  success:(AttachmentDownloadSuccess)success
+                                  failure:(AttachmentDownloadFailure)failure
 {
     self = [super init];
     if (!self) {
         return self;
     }
 
-    _networkManager = networkManager;
-
-    _attachmentPointers = @[ attachmentPointer ];
-    _attachmentIds = @[ attachmentPointer.uniqueId ];
+    _attachmentPointer = attachmentPointer;
+    _message = message;
+    _success = success;
+    _failure = failure;
 
     return self;
 }
 
-- (instancetype)initWithAttachmentProtos:(NSArray<SSKProtoAttachmentPointer *> *)attachmentProtos
-                          networkManager:(TSNetworkManager *)networkManager
-                             transaction:(YapDatabaseReadWriteTransaction *)transaction
+@end
+
+#pragma mark -
+
+@interface OWSAttachmentDownloads ()
+
+// This property should only be accessed while synchronized on this class.
+@property (nonatomic, readonly) NSMutableSet<NSString *> *downloadingAttachmentIdSet;
+// This property should only be accessed while synchronized on this class.
+@property (nonatomic, readonly) NSMutableArray<OWSAttachmentDownloadJob *> *attachmentDownloadJobQueue;
```

Comment:
I decided to use an explicit (non-durable) queue so that we could ensure that we never have two downloads for the same attachment at a time.  Maybe this was overkill?

Reply:
If we eventually wanted downloads to restart after the app is killed, we might consider porting it to the durable queue.

But this seems fine for now, as we've always previously required manual restart of failed downloads.